### ISSUE TITLE
Linking issues on AIX with OpenSSL 1.0.2 and 1.1.0 in the same location

### ISSUE THREAD

**Issue by makr** [openssl/openssl#6368]
It's probably not meant to be that way, but I wasn't aware that AIX is lacking any concept of a SONAME support and I have a decision to have OpenSSL 1.0.2 and 1.1.0 installed side-by-side, which isn't a problem on any other platform I have to look at.

The issue at hand now is, that with a linker instructions for the aix64-cc target the libssl.so.1.1 gets a dependency to libcrypto.so instead of libcrypto.so.1.1. Depending on the way I configure the side-by-side installation this either leads to no problem - if libcrypto.so points to libcrypto.so.1.1 - or to unresolved symbols - if libcrypto.so points to libcrypto.so.1.0.0 - or an error because libcrypto.so could not be found.

I partly resolved this already. For the applications that get ported to OpenSSL 1.1 I instructed the developers to use absolute paths when linking to OpenSSL 1.1, i.e. ```-brtl $(OPENSSL_LIB)/libcrypto.so.1.1 $(OPENSSL_LIB)/libssl.so.1.1``` instead of using ```-brtl -L $(OPENSSL_LIB) -lcrypto -lssl```. This is doing the right thing, as ld is removing the path component and just logs the dependency to libcrypto.so.1.1 and libssl.so.1.1 and proved to work out for executables that only need the crypto-library.

The OpenSSL build environment doesn't do such a stunt of course which results in above mentioned dependency. Therefore applications that need the SSL-library are facing unresolved symbols, because the old libcrypto.so from 1.0.2 gets loaded.

Is there a way to instruct the OpenSSL build to do a linking as I proposed to our application developers? I don't see a way to get this through into link_shlib.aix, but then I don't fully grok this mechanic anyway. When libssl.so is created the ```-lcrypto``` would need to be replaced by the absolute path to libcrypto.so.1.1...

May I also propose a name change for the libraries on AIX to be like on Windows - libcrypto-1.1.so and libssl-1.1.so - or at least have symlinks with that name?

kind regards
-- Matthias Kraft

**Comment by dot-asm** [comment]
Here is the thing. We can't possibly *find/identify* all the problems in the world. Instead we have to rely on users to tell us what the problems are. We would also very much like to hear even from OS vendors at least about their platform capabilities and at best to suggest [lasting] solution. If we don't hear from users or vendor, then you can expect all kinds of mishaps and apparently suboptimal solutions (or rather lack of solution actually suitable for target platform). AIX is apparently one of such platform. As I just ran into following. Even though aix[64]-cc targets are in the matrix, I for one have limited options to exercise them. Instead I occasionally exercise aix[64]-gcc ones. But I just had opportunity to test aix-cc, and you know what? **It doesn't pass the test!** Want to know why? Because 'cc' truncates some constants, with warning but without error, so that some of 64-bit algorithms like blake2b are effectively miscompiled. One can easily work around the problem by passing CC=xlc (at config time or to make), but question rather is how come it wasn't reported so far? Well, I can understand that people are putting together custom configs, but why not tell us that 'cc' is just inadequate choice now? [Just in case for reference, 'cc' and 'xlc' are *same* compiler, they differ only by standard compliance levels.] In other words AIX appears to be platform with no (or broken) feedback loop. So thanks for report, but don't be surprised if problem runs even deeper than your eye sees. I'll have a look at available options and suggestion, very likely pose questions...

**Comment by makr** [comment]
The first working draft patch is editing Makefile.shared as I didn't find any other fitting place. It is, however, no ideal solution as it has the ".1.1" hardwired...
```
--- Makefile.shared-original    2018-06-06 14:48:27.000000000 +0200
+++ Makefile.shared     2018-06-06 14:50:21.000000000 +0200
@@ -437,17 +437,20 @@
        NOALLSYMSFLAGS=''; \
        SHAREDFLAGS='$(CFLAGS) $(SHARED_LDFLAGS) -Wl,-bexpall,-bnolibpath,-bM:SRE'; \
        rm -f $(SHLIBNAME_FULL) 2>&1 > /dev/null ; \
+       LIBDEPS=`echo $(LIBDEPS) | sed -e 's@-L\. -lcrypto@libcrypto.so.1.1@g' -e 's@-L\. -lssl@libssl.so.1.1@g'` || :;\
        $(LINK_SO_DSO);
  link_shlib.aix:
-       @ OBJECT_MODE=`expr "x$(SHARED_LDFLAGS)" : 'x\-[a-z]*\(64\)'` || : ; \
+       @OBJECT_MODE=`expr "x$(SHARED_LDFLAGS)" : 'x\-[a-z]*\(64\)'` || : ; \
        OBJECT_MODE=$${OBJECT_MODE:-32}; export OBJECT_MODE; \
        ALLSYMSFLAGS='-bnogc'; \
        NOALLSYMSFLAGS=''; \
        SHAREDFLAGS='$(CFLAGS) $(SHARED_LDFLAGS) -Wl,-bexpall,-bnolibpath,-bM:SRE'; \
        rm -f $(SHLIBNAME_FULL) 2>&1 > /dev/null ; \
+       LIBDEPS=`echo $(LIBDEPS) | sed -e 's@-L\. -lcrypto@libcrypto.so.1.1@g' -e 's@-L\. -lssl@libssl.so.1.1@g'` || :;\
        $(LINK_SO_SHLIB_VIA_O)
  link_app.aix:
        LDFLAGS="$(CFLAGS) -Wl,-bsvr4 $(LDFLAGS)"; \
+       LIBDEPS=`echo $(LIBDEPS) | sed -e 's@-L\. -lcrypto@libcrypto.so.1.1@g' -e 's@-L\. -lssl@libssl.so.1.1@g'` || :;\
        $(LINK_APP)
  
  
```
Another catch is that the linker now complains:
```
/opt/IBM/xlC/13.1.3/bin/.orig/cc: 1501-218 (W) file libcrypto.so.1.1 contains an incorrect file suffix
```
... which makes me think that the appropriate solution would indeed be to move the version into the library name.

**Comment by levitte** [comment]
It seems like the AIX approach is to put all versions of a shared object into one library archive.

http://unix.derkeiler.com/Newsgroups/comp.unix.aix/2005-12/msg00080.html
https://gcc.gnu.org/install/specific.html#x-ibm-aix

As for that Makefile.shared hack, I'm not sure in what way it would help...

**Comment by makr** [comment]
I've read these articles, too. They describe a way to ensure backward compatibility into the last century. It is impractical from my point of view, as it requires more AIX specific code (e.g. dlopen() with RTLD_MEMBER and friends) as would be necessary.
I have been experimenting for another couple of hours and came back to the above lines. This is the way to go, if there is no other way to provide the full name of the dependencies instead of the usual "-L / -l" pair. It properly utilizes the [ld](https://www.ibm.com/support/knowledgecenter/ssw_aix_72/com.ibm.aix.cmds3/ld.htm?view=kc#ld__d7cb2a1424jani) command on AIX. Hence I'll add a PR for this issue.

The proposed changes lead to the wanted result: ```libssl.so.1.1``` then has a dependency to ```libcrypto.so.1.1``` instead of just ```libcrypto.so```. This makes it finally possible to have libcrypto.so pointing to libcrypto.so.1.0.0 and have libcrypto.so.1.1 in the same directory, too. Engines and apps are also correctly referencing libcrypto.so.1.1 and libssl.so.1.1 ...
```
bash-4.2$ dump -H libcrypto.so.1.1 libssl.so.1.1 apps/openssl engines/capi.so 

libcrypto.so.1.1:

                        ***Loader Section***
[...]

                        ***Import File Strings***
INDEX  PATH                          BASE                MEMBER              
0      /opt/.../openssl/lib:/usr/lib:/lib                                         
1                                    libpthreads.a       shr_xpg5_64.o       
2                                    libc.a              shr_64.o            

libssl.so.1.1:

                        ***Loader Section***
[...]

                        ***Import File Strings***
INDEX  PATH                          BASE                MEMBER              
0      /opt/.../openssl/lib:/usr/lib:/lib                                         
1                                    libcrypto.so.1.1                        
2                                    libc.a              shr_64.o            

apps/openssl:

                        ***Loader Section***
[...]

                        ***Import File Strings***
INDEX  PATH                          BASE                MEMBER              
0      /opt/.../openssl/lib:/usr/lib:/lib                                         
1                                    libssl.so.1.1                           
2                                    libcrypto.so.1.1                        
3                                    libpthreads.a       shr_xpg5_64.o       
4                                    libc.a              shr_64.o            

engines/capi.so:

                        ***Loader Section***
[...]

                        ***Import File Strings***
INDEX  PATH                          BASE                MEMBER              
0      /opt/.../openssl/lib:/usr/lib:/lib                                         
1                                    libcrypto.so.1.1                        
```
```
bash-4.2$ ldd -L `pwd` libcrypto.so.1.1 
libcrypto.so.1.1 needs:
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /usr/lib/libc.a(shr_64.o)
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
bash-4.2$ ldd -L `pwd` libssl.so.1.1    
libssl.so.1.1 needs:
         /.../tmp/openssl-1.1.0h-6368/libcrypto.so.1.1
         /usr/lib/libc.a(shr_64.o)
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
bash-4.2$ ldd -L `pwd` apps/openssl
apps/openssl needs:
         /.../tmp/openssl-1.1.0h-6368/libssl.so.1.1
         /.../tmp/openssl-1.1.0h-6368/libcrypto.so.1.1
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /usr/lib/libc.a(shr_64.o)
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
bash-4.2$ ldd -L `pwd` engines/capi.so 
engines/capi.so needs:
         /.../tmp/openssl-1.1.0h-6368/libcrypto.so.1.1
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /usr/lib/libc.a(shr_64.o)
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
```
```
bash-4.2$ LIBPATH=`pwd` apps/openssl version -a
OpenSSL 1.1.0h  27 Mar 2018
built on: reproducible build, date unspecified
platform: aix64-cc
options:  bn(64,64) rc4(char) des(int) blowfish(ptr) 
compiler: cc -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_MONT -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DVPAES_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/opt/.../openssl\"" -DENGINESDIR="\"/opt/.../openssl/lib/engines-1.1\"" 
OPENSSLDIR: "/opt/.../openssl"
ENGINESDIR: "/opt/.../openssl/lib/engines-1.1"
```

**Comment by makr** [comment]
Well, for the master build, $linkflags and $linklibs in libobj2shlib, obj2dso and obj2bin need to be constructed differently for this system. If $linkflags stays empty and $linklibs would contain the list of full qualified filenames for libcrypto and libssl we'd have a way.
At a first glance the code constructing '@libdirs', '$linkflags' and '$linklibs' looks almost the same in all three sub-procedures. Could it be factored out into a common sub, which could then properly handle the AIX exception?
shared-info.pl then just needs to be extended to have -bnoipath in the aix section.

**Comment by dot-asm** [comment]
> $(AR) r libcrypto$(SHLIB_EXT_SIMPLE) libcrypto$(SHLIB_EXT)

That's what IBM does with OpenSSL. I mean they seem to bundle OpenSSL with AIX and shared library is provided in /usr/lib/libcrypto.a. There actually are multiple versions. 0.9.8 and 1.0.0. I would suggest we actually follow the lead and do same. On certain level it's kind of upsetting, because they should have offered the solution...

**Comment by dot-asm** [comment]
I'd say there are two viable options to choose from (I use libcrypto alone, but it naturally applies to libssl):

1. libcrypto.a(libcrypto.so.1.1)
2. libcrypto-1.1.so

As already said I for one am leaning toward 1st. Yes, dlopen on libcrypto.so won't work (without RTLD_MEMBER and whatever it entails), but is it actually right target for dlopen? I mean internally dlopen is meant to be used primarily for loading engines and those will remain .so... Thought?

On related note, one can actually wonder if shlibloadtest that is asked to load libcrypto.so is actually a legitimate test. I mean shlibloadtest is itself linked with libcrypto.so.1.1, so what would dlopen-ing libcrypto.so actually prove? What are the expectations? That it would map *second* copy of libcrypto or just increase reference counter on libcrypto.so.1.1?

**Comment by makr** [comment]
I see. However, from a practical point of view I'd prefer the 2. option. Reason: the first needs AIX specific code _and_ AIX specific build instructions. The 2. needs _only_ AIX specific build instructions.

For reference: libcrypto.a and libssl.a as shipped by IBM contain both 32 and 64 bit DSOs. This should then be respected, too.
```
bash-4.2$ ar -Xany -tv -- /usr/lib/libcrypto.a 
rwxr-xr-x     0/0     2967588 Jul 24 15:46 2015 libcrypto.so
rwxr-xr-x     0/0     2256131 Jul 24 15:43 2015 libcrypto.so.0.9.8
rwxr-xr-x     0/0     2967588 Jul 24 15:45 2015 libcrypto.so.1.0.0
rwxr-xr-x     0/0     3376521 Jul 24 15:46 2015 libcrypto64.so
rwxr-xr-x     0/0     2606185 Jul 24 15:44 2015 libcrypto64.so.0.9.8
rwxr-xr-x     0/0     3376521 Jul 24 15:45 2015 libcrypto64.so.1.0.0
```

**Comment by dot-asm** [comment]
> the first needs AIX specific code and AIX specific build instructions

I don't follow the second assertion. The whole point with 1st option is that 3rd-party application would link in exactly same way as on any other Unix, with -lcrypto. While in second case developer would have to -lcrypto-1.1...

**Comment by makr** [comment]
OK, yes this seems to do the trick. -lcrypto would link against libcrypto.a(libcrypto64.so.1.1) for an application. That's nice for the future, so to say, but doesn't help me right now as I am to provide an environment where other developers can slowly transition their applications from 1.0.2 to 1.1... and the builds of OpenSSL 1.0.2 and 1.1.0 don't support this mechanic yet ... and I fail to see a way to patch this into the Makefile.shared/unix-Makefile.tmpl chain.
With the knowledge gained, I'd like to make a step back and start over with the topic at hand. Can anyone suggest a way how to proceed?

**Comment by makr** [comment]
I may have found a way. Pending a few more successful tests I'll update my PR accordingly...

**Comment by dot-asm** [comment]
We can naturally consider any specific input, but we can't accommodate every unconventional one of them. Instead we have to find some common ground, preferably aligned with majority's expectations [which in turn is likely to be affected by how vendor customarily does things]. But note that since there is no notion of "so-name", nothing prevents you from setting aside OpenSSL installation procedure and come up with site-specific one. I mean just don't run 'make install', but instead copy just libcrypto.so.1.1 to target location of your choice and make *your* users specify it by full path (and pass addition -b option). Of course it leaves question about openssl binary itself. Well, why not provide statically linked version on your custom location? For example you can configure with no-shared, then link libcrypto.so.1.1 yourself and copy stuff to target destination. If you bare responsibility for providing tailored environment for a hurd of application developers/maintainers, it's not unreasonable to expect that you end up with custom installation procedure anyway...

> I am to provide an environment where other developers can slowly transition their applications from 1.0.2 to 1.1...

You still end up with disjoint directories at least for headers, and in such case it's all but natural to have disjoint libraries even for link-time libraries. Question is only about run-time library location. And in such case it's sufficient if 1.0.2 and 1.1. are either a) different, or b) resolved same way IBM does things, i.e. libcrypto.a(*.so*). No?

**Comment by makr** [comment]
Unfortunately it is not that easy on this side. Sticking to _common sense_ to not emit more confusion than actually necessary is one of the goals. This is difficult enough, as every project - no matter if open or closed source - has its own _common sense_. Anyway I'll try to find similar ways across the platforms, which are perhaps backward compatible as well as future proof. It is obvious that these are in itself unreachable goals. But improvements can be made during the process and it is always an opportunity to learn.
Meanwhile I tested my idea against reality with mixed results...

It works if libcrypto.a contains libcrypto64.so.1.1 _only_.

When I took /usr/lib/libcrypto.a (provided by the system and containing 0.9.8 and 1.0.0 libraries) and added libcrypto64.so.1.1, then linked apps/openssl against this archive. The result executes, but seems to load symbols from various SOs:
```
bash-4.2$ dump -H apps/openssl

apps/openssl:

                        ***Loader Section***
                      Loader Header Information
[...]


                        ***Import File Strings***
INDEX  PATH                          BASE                MEMBER              
0      /opt/.../openssl/lib:/usr/lib:/lib                                         
1                                    libssl.a            libssl64.so         
2                                    libssl.a            libssl64.so.1.1       
3                                    libcrypto.a         libcrypto64.so      
4                                    libcrypto.a         libcrypto64.so.1.1    
5                                    libpthreads.a       shr_xpg5_64.o       
6                                    libc.a              shr_64.o            
bash-4.2$ ldd -L . apps/openssl
apps/openssl needs:
         **./libssl.a(libssl64.so)**
         ./libssl.a(libssl64.so.1.1)
         **./libcrypto.a(libcrypto64.so)**
         ./libcrypto.a(libcrypto64.so.1.1)
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /usr/lib/libc.a(shr_64.o)
         **./libcrypto.a(libcrypto64.so.1.0.0)**
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
```
Furthermore it seems that if ./libcrypto.a(libcrypto.so) _and_ ./libcrypto.so is available at link time (-L. -lcrypto), libcrypto.so is preferred. If ./libcrypto.a(libcrypto.so) and ./libcrypto.a(libcrypto.so.1.1) is available at link time, ./libcrypto.a(libcrypto.so) is preferred.

So in order to get this working, the link step link_shlib.aix in Makefile.shared needs to be amended to:

1. move away the static library, perhaps: ```mv $(STLIBNAME) $(STLIBNAME)-static```
2. create the archive with the SOs: ```$(AR) -r $(STLIBNAME) $(SHLIBNAME_FULL)```
3. move away the SOs: ```mkdir -p ./libs; mv $(SHLIBNAME) $(SHLIBNAME_FULL) ./libs/```

The 'make install'-step later on then should install the static library under a different name.

Comments? Shall I change my PR in this regard?

**Comment by dot-asm** [comment]
> Comments? Shall I change my PR in this regard?

First-most question is do we agree that libcrypto.a(libcrypto.so.1.1) is one that we choose? As for PR, it's a trick question. Problem is that it targets 1.1.0, and things look differently in 1.1.1/master. And preferred workflow is solve things in master **first**, then make assessments on what's appropriate to backport to released branches.

**Comment by dot-asm** [comment]
> This way developers will be forced to link with -L/some/where, where "developer" things reside. While users would pick corresponding libssl.a(libssl64.so.1.1) from common run-time location...

This naturally implies that applications are linked with run-time path different from link-time one. My understanding is that it can be achieved with -bsvr4 -R. I mean -bsvr4 -L/link/time/location -R/run/time/localtion -lcrypto.

**Comment by makr** [comment]
From top to bottom ...

Yes, the developer's link time library should indeed only have - say - libcrypto.a(libcrypto64.so.1.1). This is doing the right thing when linking with ```-L. -lcrypto``` then.

I'll have a look into master, too.

Although -R is "active" if -bsvr4 is given, the right command line is somewhat more complicated, as the paths given with -L would be added to the runtime search path. I think that is why OpenSSL already has -bnolibpath added. One would have to to add ```-blibpath:$(LIBRPATH):/usr/lib:/lib``` to the end of the linking step - $(LIBDEPS) in 1.1.0.

**Comment by dot-asm** [comment]
> Although -R is "active" if -bsvr4 is given, the right command line is somewhat more complicated, as the paths given with -L would be added to the runtime search path.

I can't confirm this. If I specify -bsvr4, -L is *not* added to runtime search path. Besides question is not really about what happens to apps/openssl, but what *your* users would have to do to link *their* applications.

**Comment by makr** [comment]
Sorry, I didn't read correctly. Yes, -L is added **unless** -bsvr4 is specified and with -bsvr4 specified -R might provide an easier handling as it could be specified multiple times to add components to the runtime search path.

Meanwhile I had a look into the master. libobj2shlib in unix-Makefile.tmpl could be extended. But this is out-of-scope there, actually. If I understand this correctly, libcrypto.so is now independent from libcrypto.a. So if I would enable parallel builds libcrypto.a (with shared objects on AIX) produced from libcrypto.so could be overwritten by the actual static libcrypto.a target. OR maybe the libcrypto.a  libraries should be generated in a separate directories static/ and shared/?

**Comment by makr** [comment]
Yes, it now seems to be the best option for me, too. From what I learned so far it should have only consequences for people using the dlopen() or load() families of functions. People just linking against OpenSSL should not notice. Except maybe that they may have to copy a lib###.a instead of .so around, then.

**Comment by dot-asm** [comment]
Take #6487 for a spin on your system. Do note that it doesn't touch installation yet, only compilation...

**Comment by makr** [comment]
I'll see to it tomorrow.

**Comment by makr** [comment]
Looks good.
Note, that shlibload test breaks, but this is not unexpected. I have some changes for this in above mentioned commit, but it needs further changes to the recipe.

**Comment by dot-asm** [comment]
> Looks good.

Good. #6487 is updated with install procedure and I also chose to use always pass -bsvr4 to linker. This is to enable -Wl,-R,'$(LIBRPATH)' at config time.

> Note, that shlibload test breaks, but this is not unexpected.

Of course.

> I have some changes for this in above mentioned commit, but it needs further changes to the recipe.

For now I've suppressed the test on AIX, suggestion is not handle it separately...

**Comment by dot-asm** [comment]
> For now I've suppressed the test on AIX, suggestion is not handle it separately...

I naturally meant "suggestion is to handle it separately." Related question is whether or not it should be backported to 1.1.0. Formally speaking this kind of changes are hardly appropriate even for minor release. However! Since feedback loop for platform in question seems to be broken, and since at least IBM is silently providing libcrypto.a(libcrypto.so.ver) already, one can argue that one can simply claim that "in 1.1 releases AIX is effectively supported only from 1.1.0i". This would also allow to slip cc->xlc change at least for aix-cc, i.e. 32-bit build.

**Comment by makr** [comment]
I just cherrypicked your commits and will give them a try now. I noticed one thing though: CHANGES should probably mention, that dynamic loading would now require RTLD_MEMBER with dlopen() and L_LOADMEMBER with load()/loadAndInit().

**Comment by dot-asm** [comment]
I don't quite follow, cherry-picked where? To 1.1.0? I'd be surprised if it would work. Or at least that it would work **exactly** as intended. As mentioned earlier, common workflow is do things in master, then backport to other branches. One can also view it like this, master becomes a reference to compare to. As for RTLD_MEMBER, suggestion is solve it separately.

**Comment by makr** [comment]
Installing into a pristine location works fine. Installing in a location with an existing libcrypto.a initially failed (I used ```gmake DESTDIR=/tmp/shared install``` here):
```
+ + echo libcrypto64.so.1.1;libcrypto.a
+ cut -f1 -d;
s1=libcrypto64.so.1.1
+ + echo libcrypto64.so.1.1;libcrypto.a
+ cut -f2 -d;
s2=libcrypto.a
+ + basename libcrypto64.so.1.1
fn1=libcrypto64.so.1.1
+ + basename libcrypto.a
fn2=libcrypto.a
+ :
+ a=/tmp/shared/opt/.../openssl/lib/libcrypto.a
+ echo install libcrypto64.so.1.1 -> /tmp/shared/opt/.../openssl/lib/libcrypto.a
install libcrypto64.so.1.1 -> /tmp/shared/opt/.../openssl/lib/libcrypto.a
+ [ -f /tmp/shared/opt/.../openssl/lib/libcrypto.a ]
+ trap rm -rf /tmp/ar.15400992 INT 0
+ mkdir /tmp/ar.15400992
+ cd /tmp/ar.15400992
+ cp -f /tmp/shared/opt/.../openssl/lib/libcrypto.a /tmp/shared/opt/.../openssl/lib/libcrypto.a.new
+ ar -X64 t /tmp/shared/opt/.../openssl/lib/libcrypto.a
+ ar -X64 x /tmp/shared/opt/.../openssl/lib/libcrypto.a libcrypto.so
+ strip -X32_64 -e libcrypto.so
strip: libcrypto.so -- 0654-400 Cannot open file.
        Permission denied
+ rm -rf /tmp/ar.15400992
gmake: *** [install_dev] Error 255
```
The archive's content is:
```
bash-4.2$ ar -Xany -tv /tmp/shared/opt/.../openssl/lib/libcrypto.a
r-xr-xr-x  8950/1551  4819677 Apr 04 09:09 2018 libcrypto.so
r-xr-xr-x  8950/1551  4819677 Apr 04 09:09 2018 libcrypto.so.1.0.0
```
Which is a result of the 1.0.2 installation installing the libraries with 555 mode. As I just manually packaged them into a libcrypto.a, I could take care of it. The AIX provided libraries in /usr/lib/libcrypto.a have 775 mode, which would not lead to such an error. Nevertheless, some sort of error handling or an informational error message would be nice, I think.

It is by intention that this handling is in **install_dev** and _not_ in **install_runtime**, right?

After fixing the permissions the result of installing into a location shared with OpenSSL 1.0.2 would be:
```
bash-4.2$ pwd
/tmp/shared/opt/.../openssl/lib
bash-4.2$ ls -l 
total 73608
drwxr-xr-x    2 ssxbuild ssx            4096 Apr 04 09:09 engines
drwxr-xr-x    2 ssxbuild ssx             256 Jun 18 10:27 engines-1.1
-rw-r--r--    1 ssxbuild ssx        14937601 Jun 18 10:27 libcrypto.a
lrwxrwxrwx    1 ssxbuild ssx              18 Jun 18 09:21 libcrypto.so -> libcrypto.so.1.0.0
-rwxr-xr-x    1 ssxbuild ssx         4819677 Apr 04 09:09 libcrypto.so.1.0.0
-rwxr-xr-x    1 ssxbuild ssx         5028885 Jun 18 10:27 libcrypto64.so.1.1
-rw-r--r--    1 ssxbuild ssx         6147830 Jun 18 10:27 libcrypto_a.a
-rw-r--r--    1 ssxbuild ssx         3279502 Jun 18 10:27 libssl.a
lrwxrwxrwx    1 ssxbuild ssx              15 Jun 18 09:21 libssl.so -> libssl.so.1.0.0
-rwxr-xr-x    1 ssxbuild ssx         1014605 Apr 04 09:09 libssl.so.1.0.0
-rwxr-xr-x    1 ssxbuild ssx         1184214 Jun 18 10:27 libssl64.so.1.1
-rw-r--r--    1 ssxbuild ssx         1260476 Jun 18 10:27 libssl_a.a
drwxr-xr-x    2 ssxbuild ssx             256 Apr 04 09:09 pkgconfig
bash-4.2$ ar -Xany -tv libcrypto.a
rwxr-xr-x  8950/1551  4819677 Jun 18 10:27 2018 libcrypto.so
rwxr-xr-x  8950/1551  4819677 Jun 18 10:27 2018 libcrypto.so.1.0.0
rwxr-xr-x  8950/1551  5028885 Jun 18 08:59 2018 libcrypto64.so.1.1
bash-4.2$ ldd -L . ../bin/openssl 
../bin/openssl needs:
         ./libssl.a(libssl64.so.1.1)
         ./libcrypto.a(libcrypto64.so.1.1)
         /usr/lib/libpthreads.a(shr_xpg5_64.o)
         /usr/lib/libc.a(shr_64.o)
         /unix
         /usr/lib/libcrypt.a(shr_64.o)
```

**Comment by makr** [comment]
Sorry, didn't see your comment. No, I used master as base for your commits and cherry picked into my own branch ...

**Comment by dot-asm** [comment]
> + strip -X32_64 -e libcrypto.so
> strip: libcrypto.so -- 0654-400 Cannot open file.
>        Permission denied

Ouch! In the essence one can argue both ways, both in favour of chmod u+w and against it. Against, because one can say that our installation procedure has to compatible only with itself. And it is, in sense that if you run same procedure against unique[!] location it will work. And in favour, because we made a change that requires transition, and it's only appropriate to make transition smooth. I'm leaning toward in favour...

> It is by intention that this handling is in install_dev and not in install_runtime, right?

Well, I just extended what was there to start with. In the essence it is justified question, i.e. why shared libraries are installed in install_dev and not install_runtime. I mean ideally install_dev should make arrangements that would allow 3rd-party applications to be linked. Which in most cases would mean "copy headers and create symbolic links to shared libraries [that were installed in install_runtime]." Feel free to create separate issue for this.

> ldd -L . ../bin/openssl

Do note commit message to https://github.com/openssl/openssl/pull/6487/commits/053bed665a38640e349d97fb1d565741a1f71777. I mean consider passing that additional argument at config time.

**Comment by makr** [comment]
1) After reviewing the OpenSSL 1.0.2 Makefiles I'd advise to use chmod u+w, too. It would at least be in the sense of this ticket ;-) ...

2) I'll open another ticket, if it becomes an actual issue. Otherwise I just wanted to mention it, so its an informed decision.

3) I do configure with ```-Wl,-blibpath:$(LIBRPATH):/usr/lib:/lib``` on AIX to accommodate our default product environment. On other platforms I use similar options and additionally use $ORIGIN-handling to avoid the necessity to configure the library search path using other means.

_To sum it up_: The changes as worked out so far in this ticket and distilled as code in #6487 would resolve this ticket as far as a future release line is concerned. Now for the current 1.1.0 line, I would try to condense this into a PR, if you are signaling that you are actually willing to accept such a change.

**Comment by levitte** [comment]
> I mean shlibloadtest is itself linked with libcrypto.so.1.1

Hmmm, right it does!  Via libtestutil...  that was inattentive.

**Comment by levitte** [comment]
Re `install_dev` and `install_runtime`, don't make the former depend on the latter as the latter also installs binaries, which isn't at all what users would expect if they only want to install the development files.  It's better to have a `install_shared` that they both depend on, except for mingw/cygwin, where the runtime installation installs the shared libraries in the bin directory instead of the lib directory (because that's what Windows wants).

**Comment by dot-asm** [comment]
#6487 is merged. Question about if it should be back-ported to 1.1.0 remains open. As mentioned above AIX turned to have broken feedback loop, so that one can claim that AIX is not really-really-fully supported in 1.1.0, and firmly advice people to go straight for 1.1.1. It's not over-night change to start with, so 1.1.1 being not released but being "around corner" should not pose problem. So I for one kind of leaning toward not back-porting, but @makr is likely to disagree.

**Comment by richsalz** [comment]
This is no longer a 1.1.1 issue, so removing that tag.
I suggest that unless someone steps up with PR's for 1.1.0/1.0.2, we close this.

**Comment by makr** [comment]
I am not going to backport the solution and its solved in 1.1.1. Closing as per comments from @dot-asm and @richsalz .
